R/aalen_johansen.R

```diff
@@ -0,0 +1,114 @@
+#' Aalen Johansen Estimator
+#'
+#' This function calculates the Aalen Johansen estimator of
+#' adverse events  observed in `[0, tau]`.
+#' Please also refer to  \insertCite{stegherr_meta_analytic_2021;textual}{savvyr}.
+#'
+#' Estimating and comparing adverse event probabilities in the presence of varying follow-up times and competing events
+#' Regina Stegherr, Claudia Schmoor, Michael Lübbert, Tim Friede, Jan Beyersmann
+#' First published: 18 May 2021 https://doi.org/10.1002/pst.2130
+#'
+#' @typed data: data.frame
+#'  with columns including
+#'  - `time_to_event`: Time to the first AE, death or soft competing event.
+#'  - `type_of_event`: 0 for censored, 1 for AE, 2 for death, 3 for soft competing event.
+#'
+#' @typed tau: number
+#'  milestone at which Aalen-Johansen is computed.
+#'
+#' @typedreturn vector
+#'  with the following entries:
+#'
+#' - `ae_prob`: Estimated probability of AE.
+#' - `ae_prob_var`: Variance of that estimate.
+#' - `ce_prob`: Estimated probability of competing events.
+#' - `ce_prob_var`: Variance of competing events.
+#'
+#' @export
+#'
+#' @references
+#' \insertRef{stegherr_meta_analytic_2021}{savvyr}
+#'
+#' @examples
+#' set.seed(123)
+#' dat <- generate_data(n = 5, cens = c(2, 5), haz_ae = 2, haz_death = 3, haz_soft = 5)
+#' one_minus_kaplan_meier(dat, tau = 4)
+aalen_johansen <- function(data,
+                           ce,
+                           tau) {
+  assert_data_frame(data, any.missing = FALSE, min.rows = 1, min.cols = 2)
+  assert_numeric(data$time_to_event, lower = 0, finite = TRUE)
+  assert_integerish(data$type_of_event, any.missing = FALSE)
+  assert_subset(data$type_of_event, c(0, 1, 2, 3))
+  assert_number(tau, finite = TRUE)
+  assert_true(tau > 0)
+  assert_subset(ce, c(2, 3))
+
+  data$type_of_event_accounted <- ifelse(ce == 2 & data$type_of_event == 3, 0,
+    ifelse(ce == 3 & data$type_of_event == 3, 2, data$type_of_event)
+  )
+
+  time <- data$time_to_event
+  type2 <- data$type_of_event_accounted
+
+  # conditions
+  c1 <- sum(data$type_of_event_accounted == 1)
+  c2 <- sum(data$type_of_event_accounted == 2)
+
+  if (c1 == 0) {
+    ae_prob <- 0
+    ae_prob_var <- 0
+  }
+
+  if (c2 == 0) {
+    ce_prob <- 0
+    ce_prob_var <- 0
+  }
+
+  # define auxiliary objects
+  help <- data.frame(id = data$id)
+  help$from <- 0
+  help$time <- ifelse(time == 0, 0.001, time)
+  tra <- matrix(FALSE, 2, 2)
+  tra[1, 2] <- TRUE
+  state.names <- as.character(0:1)
+
+  if (c1 == 0 & c2 != 0) {
+    help$to <- ifelse(type2 != 2, "cens", type2 - 1)
+    etmmm <- etm::etm(help, state.names, tra, "cens", s = 0)
+    setmm <- summary(etmmm)[[2]]
+    ce_prob <- setmm[sum(setmm$time <= tau), ]$P
+    ce_prob_var <- setmm[sum(setmm$time <= tau), ]$var
```

Comment:
to avoid repetition, first save `setmm[sum(setmm$time <= tau), ` in a variable and then access elements

Reply:
I would leave this as its repeated only once?

Reply:
yeah but still, it will be also easier to read then

Reply:
changed